File: 29Items-Projects/7Algorithms/6-Backtracking/java/src/generateparentheses/GenerateParentheses.java
package generateparentheses;

import java.util.*;

public class GenerateParentheses {

    private final int n;
    private final List<String> solutions = new ArrayList<>();

    public GenerateParentheses(int n) {
        if (n <= 0) throw new IllegalArgumentException("n must be > 0");
        this.n = n;
    }

    public void solveAndPrint() {
        backtrack(new StringBuilder(), 0, 0);
        System.out.printf("Found %d valid combination(s) for n=%d%n%n", solutions.size(), n);

        int limit = Math.min(solutions.size(), 8);
        for (int i = 0; i < limit; i++) {
            System.out.printf("#%d: %s%n", i + 1, solutions.get(i));
        }
        if (solutions.size() > limit) {
            System.out.printf("... and %d more.%n", solutions.size() - limit);
        }
    }

    private void backtrack(StringBuilder current, int open, int close) {
        if (current.length() == n * 2) {
            solutions.add(current.toString());
            return;
        }
        if (open < n) {
            current.append('(');
            backtrack(current, open + 1, close);
            current.deleteCharAt(current.length() - 1);
        }
        if (close < open) {
            current.append(')');
            backtrack(current, open, close + 1);
            current.deleteCharAt(current.length() - 1);
        }
    }
}

File: 29Items-Projects/7Algorithms/6-Backtracking/python/generate_parentheses.py
"""Generate Parentheses — all valid combinations of n pairs of parentheses."""


class GenerateParentheses:
    def __init__(self, n: int):
        if n <= 0:
            raise ValueError("n must be > 0")
        self.n = n
        self.solutions: list[str] = []

    def solve_and_print(self) -> None:
        self._backtrack("", 0, 0)
        print(f"Found {len(self.solutions)} valid combination(s) for n={self.n}\n")

        limit = min(len(self.solutions), 8)
        for i, sol in enumerate(self.solutions[:limit], 1):
            print(f"#{i}: {sol}")
        if len(self.solutions) > limit:
            print(f"... and {len(self.solutions) - limit} more.")

    def _backtrack(self, current: str, open_cnt: int, close_cnt: int) -> None:
        if len(current) == self.n * 2:
            self.solutions.append(current)
            return
        if open_cnt < self.n:
            self._backtrack(current + "(", open_cnt + 1, close_cnt)
        if close_cnt < open_cnt:
            self._backtrack(current + ")", open_cnt, close_cnt + 1)
